src/jit/optcse.cpp

```diff
@@ -920,79 +968,251 @@ void Compiler::optValnumCSE_Availablity()
         GenTree* stmt;
         GenTree* tree;
 
-        /* Make the block publicly available */
+        // Make the block publicly available
 
         compCurBB = block;
 
+        // Retrieve the available CSE's at the start of this block
+
         BitVecOps::Assign(cseTraits, available_cses, block->bbCseIn);
 
         optCSEweight = block->getBBWeight(this);
 
-        /* Walk the statement trees in this basic block */
+        // Walk the statement trees in this basic block
 
         for (stmt = block->FirstNonPhiDef(); stmt; stmt = stmt->gtNext)
         {
             noway_assert(stmt->gtOper == GT_STMT);
 
-            /* We walk the tree in the forwards direction (bottom up) */
+            // We walk the tree in the forwards direction (bottom up)
+
             for (tree = stmt->gtStmt.gtStmtList; tree; tree = tree->gtNext)
             {
                 if (IS_CSE_INDEX(tree->gtCSEnum))
                 {
-                    unsigned int cseBit = genCSEnum2bit(tree->gtCSEnum);
-                    CSEdsc*      desc   = optCSEfindDsc(tree->gtCSEnum);
+                    unsigned     CSEnum = GET_CSE_INDEX(tree->gtCSEnum);
```

Comment:
Seems like somewhere in here (or perhaps in the method header) you should outline the general approach: the intention is to determine for each candidate if the intersection of all the exceptions provided by the defs is a superset of the union of the exceptions expected by the (qualified) uses.

The algorithm uncovers the defs and uses gradually and so incrementally building up both the intersection def set and the union use set.

It would be good to make sure we have (or can point at) test cases that show the various cases of this code getting covered.

I am still thinking about it but wonder if there is some implicit priority in choosing uses that should be thought about more carefully. Suppose we first find a def with exceptions 1, 2, 3 and then a use with exceptions 1, 2, 3, and then a def with exceptions 2, 3 and then a use with exceptions 2, 3.  By the current algorithm we would not do any CSE here. But we could still CSE the 2,3 use.

So I wonder if instead we should not disqualify any uses until we've seen them all, and then pick the subset of uses that is compatible with what the definitions can provide.

Reply:
Typically expressions with the same ValueNum generate exactly the same exception sets.
There are two way that we can get different exception sets with the same value number.

1. We used an arithmetic identiity:  
e.g.:  `(p.a + q.b) * 0`  -- The normal value for the expression is set to zero because of the multiply, 
e.g.   `(p.a - p.a )`        -- The normal value for the expression is set to zero because of the subtraction

2. We stored an expression into a LclVar or into Memory and used it later in a CSE:
e.g.   't = p.a;   e1=(t + q.b);  e2=(p.a+q.b)'  --  e1 has one NullPtrExc and e2 has two.
e.g.  `m.a = p.a;  e1=(m.a + q.b); e2=(p.a + q.b)  -- e1 and e2 have different exception sets.

The bug cases were due to case 1.  
The small code regressions are due to case 2.

I believe that case 2 can be fixed by adding excepotion sets to the ValueNum that we use to track the Global Heap.  After each operation we would update the Global Heap with any new NullPtrExc or other exception sets.  Then the CSE uses could also rely upon exception sets tracked by the Global Heap.

Reply:
Added large  comment addressing this to the method header

Reply:
Thanks for the comment -- my main concern was not how this can happen (though it is nice to now have specifics here), but how often it can happen. 

As long as it's "rare" that different CSE defs have different exception sets then it doesn't matter much whether you disqualify uses eagerly or wait until you've seen them all, as it is unlikely you'll ever hit the case I describe where you miss CSEing some use because you saw some other more constraining use first.

But maybe it's worth mentioning that the approach you take here may not find the "maximal" set of CSEs in some rare cases.